addon/utils/merge-deep.js

```diff
@@ -137,6 +137,10 @@ export default function mergeDeep(target, source, options = {}) {
   } else if (sourceIsArray) {
     return source;
   } else {
-    return mergeTargetAndSource(target, source, options);
+    try {
+      return mergeTargetAndSource(target, source, options);
+    } catch(e) {
+      throw new Error('Unable to `mergeDeep` with your data');
+    }
```

Comment:
As a code-golf, maybe you can drop `else` clause here? As it is the last code path?

For the error message, do you think - maybe it worth to specify that it is due to the merging of two ember-data models, that result in circular merge sequence?